engines/pisa-0.8.1/tools/do_query.cpp

```diff
@@ -0,0 +1,108 @@
+#include <iostream>
+
+#include <boost/algorithm/string.hpp>
+#include <cursor/block_max_scored_cursor.hpp>
+#include <cursor/cursor.hpp>
+#include <cursor/scored_cursor.hpp>
+#include <index_types.hpp>
+#include <mio/mmap.hpp>
+#include <query/algorithm.hpp>
+#include <query/queries.hpp>
+#include <query/term_processor.hpp>
+#include <scorer/scorer.hpp>
+#include <spdlog/sinks/stdout_color_sinks.h>
+#include <spdlog/spdlog.h>
+#include <topk_queue.hpp>
+#include <wand_data.hpp>
+#include <wand_data_compressed.hpp>
+
+int main(int argc, char const *argv[]) {
+  using namespace pisa;
+  spdlog::drop("");
+  spdlog::set_default_logger(spdlog::stderr_color_mt(""));
+
+  std::string terms_file = "termlex";
+  std::string wand_data_filename = "wand";
+  std::string index_filename = "index";
+
+  std::string stemmer = "porter2";
+  std::string scorer_name = "bm25";
+  size_t k = 10;
+  bool intersection = false;
+
+  auto term_processor = TermProcessor(terms_file, std::nullopt, stemmer);
+  using wand_uniform_index_quantized =
+      wand_data<wand_data_compressed<PayloadType::Quantized>>;
+  wand_uniform_index_quantized wdata;
+
+  mio::mmap_source md;
+  std::error_code error;
+  md.map(wand_data_filename, error);
+  if (error) {
+    std::cerr << "error mapping file: " << error.message() << ", exiting..."
+              << std::endl;
+    throw std::runtime_error("Error opening file");
+  }
+  mapper::map(wdata, md, mapper::map_flags::warmup);
+
+  using IndexType = block_simdbp_index;
+
+  IndexType index;
+  spdlog::info("Loading index from {}", index_filename);
+  mio::mmap_source m(index_filename.c_str());
+  mapper::map(index, m);
+
+  auto scorer = scorer::from_name(scorer_name, wdata);
+
+  std::string line;
+  while (std::getline(std::cin, line)) {
+    size_t count = 0;
+    std::vector<std::string> tokens;
+    boost::split(tokens, line, boost::is_any_of("\t"));
+    if (boost::starts_with(tokens[1], "\"")) {
+      std::cout << "UNSUPPORTED\n";
+      continue;
+    } else if (boost::starts_with(tokens[1], "+")) {
+      intersection = true;
```

Comment:
i think this is ok right now... but we may add boolean queries mixing intersection and union in the future.

Reply:
No problem